### PR TITLE
Always print parentheses around arrow function parameters

### LINKED ISSUE #13198: [Bug]: Edge-case exception while printing single-parameter arrow function
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```js
const { parse } = require("@babel/parser");
const t = require("@babel/types");
const { default: generate } = require("@babel/generator");

// Create an identifier with a location and a leading comment.
const param = parse("(/*x*/ a)").program.body[0].expression;

// Create an async arrow function without a location.
const ast = t.arrowFunctionExpression(
    [param],
    t.identifier("a"),
    /* async */ true,
);

// TypeError: Cannot read property 'start' of undefined
generate(ast);
```

### Configuration file name

_No response_

### Configuration

_No response_

### Current and expected behavior

Expected: Babel generates valid output source for the AST.

Actual:
```
$ yarn node repro.js 
/Users/user/babel/packages/babel-generator/lib/generators/methods.js:137
      if (firstParam.loc && firstParam.loc.start.line > node.loc.start.line) {
                                                                 ^

TypeError: Cannot read property 'start' of undefined
    at Generator.ArrowFunctionExpression (/Users/user/babel/packages/babel-generator/lib/generators/methods.js:137:66)
    at /Users/user/babel/packages/babel-generator/lib/printer.js:287:19
    at Buffer.withSource (/Users/user/babel/packages/babel-generator/lib/buffer.js:188:28)
    at Generator.withSource (/Users/user/babel/packages/babel-generator/lib/printer.js:153:15)
    at Generator.print (/Users/user/babel/packages/babel-generator/lib/printer.js:286:10)
    at Generator.generate (/Users/user/babel/packages/babel-generator/lib/printer.js:45:10)
    at Generator.generate (/Users/user/babel/packages/babel-generator/lib/index.js:26:18)
    at generate (/Users/user/babel/packages/babel-generator/lib/index.js:100:14)
    at Object.<anonymous> (/Users/user/babel/repro.js:8:1)
    at Module._compile (node:internal/modules/cjs/loader:1092:14)
```

### Environment

- Current Babel main (fb908fd87e1bf8a5ebe2c881c0baab1488c917e7)

### Possible solution

Originally I was planning to PR an additional null check, but on closer inspection (and in light of #13136) I think it makes more sense to get rid of all this special-case code and just delegate to the regular param-printing code.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Zalathar**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #13198  
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Existing tests updated
| Documentation PR Link    | 
| Any Dependency Changes?  | No
| License                  | MIT

Some arrow functions can safely be printed without parentheses around a single identifier parameter.

However, there are many factors that can make this unsafe (e.g. type annotations, comments), and several generator features that can silently introduce edge cases (`retainLines`, `auxiliaryCommentBefore`/`auxiliaryCommentAfter`). These make it hard to be confident that printing the bare identifier really is safe.

This PR therefore makes the generator always print the parentheses, even when they aren't strictly necessary, greatly simplifying the implementation.

### CODE REVIEW

**Comment by Zalathar** [comment]
I have an alternate version of this change that tries to retain no-parens in some cases, and instead just tries to make the logic simpler and more robust.

But I like the idea of just always spending two characters to guarantee correct output.

**Review by nicolo-ribaudo — COMMENTED** [review]
Looking at the output, maybe we could still keep the single param, no async, no comments case (which is one of the most common ones) parens-less, or at least do it when the `compact: true` option is set.

It should still be safe, since it doesn't need to check loc and accidental line breaks don't cause problems.

File: packages/babel-generator/test/fixtures/flow/predicates/output.js [hunk 1]
```diff
@@ -14,4 +14,4 @@ var f = (x: mixed): %checks => typeof x === "string";
 
 const foo2 = (x: mixed): boolean %checks => typeof x === "string";
 
-x: %checks => x !== null;
\ No newline at end of file
+(x): %checks => x !== null;
```
Comment (nicolo-ribaudo) [inline packages/babel-generator/test/fixtures/flow/predicates/output.js:17]: Oh this was a bug

**Comment by Zalathar** [comment]
OK, I'll polish up the version that tries to keep the no-parens in simple cases, and see what that looks like.

**Comment by Zalathar** [comment]
I opened a separate PR for the alternate version at #13204, to avoid having to force-push back and forth between the two versions.

**Comment by Zalathar** [comment]
Closed in favour of #13204.

**Pull request closed without merging**